### DISCUSSION TITLE
Request for removing `impl Module for Box<dyn Module>`

### DISCUSSION THREAD

**Discussion by hyf0** [web-infra-dev/rspack#2762] [Core] [answered]
# tl;dr

`impl Module for Box<dyn Module>`  make `&Box<dyn Module>` look like a `&dyn Module`, but they are totally different things. People easily fall into traps.

## details

I have seen some people debugging for not realizing `&Box<dyn T>` and `&dyn T` are different.

And that's the case I meet:

`NormalModule` overwrote the `name_for_condition` method of `Module`

https://github.com/web-infra-dev/rspack/blob/2c4c1daf81d388cd4fc0b54049d9e2a762e34418/crates/rspack_core/src/normal_module.rs#L636-L645

You put the `NormalModule` in the `Box` and erase the type `Box` and you get `Box<dyn Module>`.

`Box<dyn Module>` also implements the `Module` trait but forgets to redirect the `name_for_condition` method to the original `NormalModule`.

The re-direct code looks like 

https://github.com/web-infra-dev/rspack/blob/2c4c1daf81d388cd4fc0b54049d9e2a762e34418/crates/rspack_core/src/module.rs#L155-L158

The default implementation for `name_for_condition` method of `Module` is

https://github.com/web-infra-dev/rspack/blob/2c4c1daf81d388cd4fc0b54049d9e2a762e34418/crates/rspack_core/src/module.rs#L93-L97

So, when you call `name_for_condition()` in `&Box<dyn Module>` you would always get a `None`.

## traps

You will fall into traps:

- If you are adding a new method in `Module` but forgetting to redirect in `impl Module for Box<dyn Module>`
- If you forgot to redirect existing methods of `Module` in `impl Module for Box<dyn Module>`

## solution

I have talked about this with @h-a-n-a before, the idea of `impl Module for Box<dyn Module>` is to reduce boilerplate code for deref `Box`.

I wonder whether there are other solutions to achieve the same goal. After investigating the code, I found most of the problems are due to using `&BoxModule` as params or return values.

Take this code as example

https://github.com/web-infra-dev/rspack/blob/2c4c1daf81d388cd4fc0b54049d9e2a762e34418/crates/rspack_core/src/module_graph/mod.rs#L162-L165

There's no need for returning `Box<&dyn Module>` due to no use cases(so far) caring about the `Box` in `Box<&dyn Module>`. Returning `Option<&dyn Module>` would be enough.

The argument also applies to using `&BoxModule` in function parameters. So, we just need to replace all `&BoxModule` with `&dyn Module`, which would also achieve the goal: reduce boilerplate code for deref `Box` and avoid the traps mentioned above.

**Comment by IWANABETHATGUY** [comment]
For this question: 
![image](https://user-images.githubusercontent.com/17974631/232430332-417ce431-30cb-49a0-90be-5c8c507ebb78.png)
What about removing the default implementation?

**Reply by hyf0** [reply]
If that's the case we need to remove all default implementations to get checking from the compiler.

**Comment by hyf0** [comment, marked as answer]
sync up:

We would remove `impl Module for Box<dyn Module>` in the future but I don't have time now. I would create a good first issue 
in case anyone wants to do this.
